analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/TypeCatalog.cs

```diff
@@ -0,0 +1,43 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn;
+
+internal class TypeCatalog
+{
+    private readonly Compilation compilation;
+    private readonly Dictionary<string, INamedTypeSymbol> cache = new();
+
+    public INamedTypeSymbol SystemIndexOutOfRangeException => Get("System.IndexOutOfRangeException");
+    public INamedTypeSymbol SystemNullReferenceException => Get("System.NullReferenceException");
+    public INamedTypeSymbol SystemInvalidCastException => Get("System.InvalidCastException");
+
+    public TypeCatalog(Compilation compilation)
+    {
+        this.compilation = compilation;
+    }
+
+    private INamedTypeSymbol Get(string typeName) =>
+        cache.GetOrAdd(typeName, name => compilation.GetTypeByMetadataName(name));
+}
```

Comment:
Let's do it like this. I did some measurements and it comes out 20x faster than the dictionary. (Calling the compilation was counted together with usages to be fair)
```suggestion
internal class TypeCatalog
{
    public INamedTypeSymbol SystemIndexOutOfRangeException { get; }
    public INamedTypeSymbol SystemNullReferenceException { get; }
    public INamedTypeSymbol SystemInvalidCastException { get; }

    public TypeCatalog(Compilation compilation)
    {
        SystemIndexOutOfRangeException = compilation.GetTypeByMetadataName("System.IndexOutOfRangeException");
        SystemNullReferenceException = compilation.GetTypeByMetadataName("System.NullReferenceException");
        SystemInvalidCastException = compilation.GetTypeByMetadataName("System.InvalidCastException");
    }
}
```